src/repositories/crates/api/v1/mod.rs

```diff
@@ -20,25 +23,44 @@ async fn download(
         if let Some(crate_version) = info.versions.iter().find(|e| e.num == version) {
             let link = &crate_version.dl_path;
 
-            if let Ok(mut upstream) = policy
-                .client
-                .get(format!("https://crates.io/{link}"))
-                .send()
-                .await
-            {
+            let url = format!("https://crates.io/{link}");
+            if let Ok(mut upstream) = policy.client.get(url.clone()).send().await {
                 if let Ok(payload) = upstream.body().limit(20_000_000).await {
                     let digest = sha256::digest(payload.as_ref());
                     let uuids = search(digest.clone()).await;
                     println!("{crate_name} {version} = {digest} {uuids:?}");
 
-                    let mut response = HttpResponse::Ok();
-                    if let Some(v) = upstream.headers().get(header::CONTENT_TYPE) {
-                        response.append_header((header::CONTENT_TYPE, v));
+                    let id = ArtifactIdentifier::Crate {
+                        name: crate_name.clone(),
+                    };
+                    let context = Context::new(
+                        url,
+                        sha256::digest(payload.as_ref()), // todo: double check this
+                        id,
+                        crates.scope.to_owned(),
+                    );
+
+                    match policy.evaluate(&context).await {
+                        Ok(None) => {
+                            log::info!("Policy evaluation success");
+                            let mut response = HttpResponse::Ok();
+                            if let Some(v) = upstream.headers().get(header::CONTENT_TYPE) {
+                                response.append_header((header::CONTENT_TYPE, v));
+                            }
+                            let disposition =
+                                format!("attachment; filename=\"{crate_name}-{version}.crate\"");
+                            response.append_header((header::CONTENT_DISPOSITION, disposition));
```

Comment:
I'm not sure about this. Are those two headers enough? Are they what the client expects? In the Maven handler, I just relay all the headers and use the upstream status. If we think that's correct, maybe we should factor it out to a common lib. Thoughts?

```suggestion
                        let mut response = HttpResponseBuilder::new(upstream.status());
                        for header in upstream.headers().iter() {
                            response.insert_header(header);
                        }
                        response.body(payload)
```

Reply:
@jcrossley3 I had actually just moved the existing code into that block if the policy eval succeeds, but I agree it could use a double check and also that we should implement common libraries where possible (that's going to be one of the next things I want to look at more generally). I like your suggestion and can update.